File: frontend/src/services/masterDataService.js
import api from './api';

const MASTER_DATA_ENDPOINTS = {
  BY_TYPE: (dataType) => `/api/master-data/type/${dataType}`,
  ALL_TYPES: '/api/master-data/types',
  CATEGORIES: '/api/master-data/categories',
  SKILL_LEVELS: '/api/master-data/skill-levels',
  ADMIN_ALL: '/api/master-data/admin',
  ADMIN_BY_TYPE: (dataType) => `/api/master-data/admin/type/${dataType}`,
  ADMIN_CREATE: '/api/master-data/admin',
  ADMIN_UPDATE: (id) => `/api/master-data/admin/${id}`,
  ADMIN_DELETE: (id) => `/api/master-data/admin/${id}`,
  ADMIN_SEED: '/api/master-data/admin/seed',
};

// Data type constants
export const DATA_TYPES = {
  COURSE_CATEGORY: 'COURSE_CATEGORY',
  SKILL_LEVEL: 'SKILL_LEVEL',
  DURATION_TYPE: 'DURATION_TYPE',
  ENROLLMENT_STATUS: 'ENROLLMENT_STATUS',
};

// Cache for master data
const cache = new Map();
const CACHE_DURATION = 5 * 60 * 1000; // 5 minutes

/**
 * Get master data by type (with caching)
 */
export async function getByType(dataType) {
  const cacheKey = `type:${dataType}`;
  const cached = cache.get(cacheKey);
  const now = Date.now();

  if (cached && (now - cached.time) < CACHE_DURATION) {
    return cached.data;
  }

  try {
    const response = await api.get(MASTER_DATA_ENDPOINTS.BY_TYPE(dataType));
    cache.set(cacheKey, { data: response.data, time: now });
    return response.data;
  } catch (error) {
    console.error(`Failed to load master data for type ${dataType}:`, error);
    return [];
  }
}

/**
 * Get course categories
 */
export async function getCourseCategories() {
  return getByType(DATA_TYPES.COURSE_CATEGORY);
}

/**
 * Get skill levels
 */
export async function getSkillLevels() {
  return getByType(DATA_TYPES.SKILL_LEVEL);
}

/**
 * Get all available data types
 */
export async function getAllTypes() {
  const response = await api.get(MASTER_DATA_ENDPOINTS.ALL_TYPES);
  return response.data;
}

/**
 * Clear cache
 */
export function clearCache() {
  cache.clear();
}

// ========== Admin functions ==========

/**
 * Get all master data (admin)
 */
export async function getAllMasterData() {
  const response = await api.get(MASTER_DATA_ENDPOINTS.ADMIN_ALL);
  return response.data;
}

/**
 * Get master data by type including inactive (admin)
 */
export async function getByTypeAdmin(dataType) {
  const response = await api.get(MASTER_DATA_ENDPOINTS.ADMIN_BY_TYPE(dataType));
  return response.data;
}

/**
 * Create a new master data entry
 */
export async function createMasterData(data) {
  const response = await api.post(MASTER_DATA_ENDPOINTS.ADMIN_CREATE, data);
  clearCache();
  return response.data;
}

/**
 * Update a master data entry
 */
export async function updateMasterData(id, data) {
  const response = await api.put(MASTER_DATA_ENDPOINTS.ADMIN_UPDATE(id), data);
  clearCache();
  return response.data;
}

/**
 * Delete a master data entry
 */
export async function deleteMasterData(id) {
  await api.delete(MASTER_DATA_ENDPOINTS.ADMIN_DELETE(id));
  clearCache();
}

/**
 * Seed default master data
 */
export async function seedDefaults() {
  const response = await api.post(MASTER_DATA_ENDPOINTS.ADMIN_SEED);
  clearCache();
  return response.data;
}

export default {
  getByType,
  getCourseCategories,
  getSkillLevels,
  getAllTypes,
  clearCache,
  getAllMasterData,
  getByTypeAdmin,
  createMasterData,
  updateMasterData,
  deleteMasterData,
  seedDefaults,
  DATA_TYPES,
};

File: src/main/java/com/uyirgene/config/MasterDataService.java
package com.uyirgene.config;

import com.uyirgene.exception.EntityNotFoundException;
import lombok.RequiredArgsConstructor;
import org.springframework.cache.annotation.CacheEvict;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class MasterDataService {

    private final MasterDataRepository repository;

    // Common data types
    public static final String TYPE_COURSE_CATEGORY = "COURSE_CATEGORY";
    public static final String TYPE_SKILL_LEVEL = "SKILL_LEVEL";
    public static final String TYPE_DURATION_TYPE = "DURATION_TYPE";
    public static final String TYPE_ENROLLMENT_STATUS = "ENROLLMENT_STATUS";

    /**
     * Get all master data entries (for admin)
     */
    public List<MasterData> getAllMasterData() {
        return repository.findAll();
    }

    /**
     * Get active master data by type
     */
    @Cacheable(value = "masterData", key = "#dataType")
    public List<MasterData> getByType(String dataType) {
        return repository.findByDataTypeAndActiveTrueOrderByDisplayOrderAsc(dataType);
    }

    /**
     * Get all master data by type (including inactive, for admin)
     */
    public List<MasterData> getByTypeAdmin(String dataType) {
        return repository.findByDataTypeOrderByDisplayOrderAsc(dataType);
    }

    /**
     * Get master data by type and code
     */
    public MasterData getByTypeAndCode(String dataType, String code) {
        return repository.findByDataTypeAndCode(dataType, code)
                .orElseThrow(() -> new EntityNotFoundException("Master data not found"));
    }

    /**
     * Get child master data entries
     */
    public List<MasterData> getChildren(Long parentId) {
        return repository.findByParentIdOrderByDisplayOrderAsc(parentId);
    }

    /**
     * Get all distinct data types
     */
    public List<String> getAllDataTypes() {
        return repository.findAll().stream()
                .map(MasterData::getDataType)
                .distinct()
                .sorted()
                .collect(Collectors.toList());
    }

    /**
     * Create a new master data entry
     */
    @Transactional
    @CacheEvict(value = "masterData", key = "#masterData.dataType")
    public MasterData create(MasterData masterData) {
        if (repository.existsByDataTypeAndCode(masterData.getDataType(), masterData.getCode())) {
            throw new IllegalArgumentException("Entry with type '" + masterData.getDataType() + "' and code '" + masterData.getCode() + "' already exists");
        }
        return repository.save(masterData);
    }

    /**
     * Update an existing master data entry
     */
    @Transactional
    @CacheEvict(value = "masterData", allEntries = true)
    public MasterData update(Long id, MasterData updates) {
        MasterData data = repository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Master data not found"));

        if (updates.getLabel() != null) data.setLabel(updates.getLabel());
        if (updates.getDescription() != null) data.setDescription(updates.getDescription());
        if (updates.getIconUrl() != null) data.setIconUrl(updates.getIconUrl());
        if (updates.getDisplayOrder() != null) data.setDisplayOrder(updates.getDisplayOrder());
        if (updates.getActive() != null) data.setActive(updates.getActive());
        if (updates.getMetadata() != null) data.setMetadata(updates.getMetadata());

        return repository.save(data);
    }

    /**
     * Delete a master data entry
     */
    @Transactional
    @CacheEvict(value = "masterData", allEntries = true)
    public void delete(Long id) {
        if (!repository.existsById(id)) {
            throw new EntityNotFoundException("Master data not found");
        }
        repository.deleteById(id);
    }

    /**
     * Seed default master data
     */
    @Transactional
    @CacheEvict(value = "masterData", allEntries = true)
    public void seedDefaultData() {
        // Course Categories
        seedCategory(TYPE_COURSE_CATEGORY, "PROGRAMMING", "Programming", 1);
        seedCategory(TYPE_COURSE_CATEGORY, "WEB_DEVELOPMENT", "Web Development", 2);
        seedCategory(TYPE_COURSE_CATEGORY, "MOBILE_DEVELOPMENT", "Mobile Development", 3);
        seedCategory(TYPE_COURSE_CATEGORY, "DATA_SCIENCE", "Data Science", 4);
        seedCategory(TYPE_COURSE_CATEGORY, "MACHINE_LEARNING", "Machine Learning", 5);
        seedCategory(TYPE_COURSE_CATEGORY, "FOOD_SAFETY", "Food Safety", 6);
        seedCategory(TYPE_COURSE_CATEGORY, "QUALITY_MANAGEMENT", "Quality Management", 7);
        seedCategory(TYPE_COURSE_CATEGORY, "ISO_STANDARDS", "ISO Standards", 8);
        seedCategory(TYPE_COURSE_CATEGORY, "MICROBIOLOGY", "Microbiology", 9);
        seedCategory(TYPE_COURSE_CATEGORY, "REGULATORY", "Regulatory Compliance", 10);
        seedCategory(TYPE_COURSE_CATEGORY, "BUSINESS", "Business", 11);
        seedCategory(TYPE_COURSE_CATEGORY, "DESIGN", "Design", 12);
        seedCategory(TYPE_COURSE_CATEGORY, "OTHER", "Other", 99);

        // Skill Levels
        seedCategory(TYPE_SKILL_LEVEL, "BEGINNER", "Beginner", 1);
        seedCategory(TYPE_SKILL_LEVEL, "INTERMEDIATE", "Intermediate", 2);
        seedCategory(TYPE_SKILL_LEVEL, "ADVANCED", "Advanced", 3);
        seedCategory(TYPE_SKILL_LEVEL, "EXPERT", "Expert", 4);

        // Duration Types
        seedCategory(TYPE_DURATION_TYPE, "SELF_PACED", "Self-Paced", 1);
        seedCategory(TYPE_DURATION_TYPE, "FIXED_SCHEDULE", "Fixed Schedule", 2);
        seedCategory(TYPE_DURATION_TYPE, "LIVE_CLASSES", "Live Classes", 3);
    }

    private void seedCategory(String dataType, String code, String label, int order) {
        if (!repository.existsByDataTypeAndCode(dataType, code)) {
            repository.save(MasterData.builder()
                    .dataType(dataType)
                    .code(code)
                    .label(label)
                    .displayOrder(order)
                    .active(true)
                    .build());
        }
    }
}
